### DISCUSSION TITLE
Save/restore compressed transitions

### DISCUSSION THREAD

**Discussion by keiohta** [ymd-h/cpprb#10] [Feature Request]
I'd like to request a feature that enables users to save and restore stored transitions with as small data size as possible.
The current implementation does allow users to do that as:
```python
samples = rb.get_all_transitions()
joblib.dump(samples, "filename.pkl", compress=3)
```
, but this can be improved.

For example, in most cases `samples["states"]` and `samples["next_states"]` considerably overlaps except the last state, and since the state occupies the most part of the stored data (imagine Atari games whose input is 84x84x4 and output is scalar value), so just removing the overlapped state enables  to reduce almost half of total data size.

There should be more sophisticated ways to enable that, so let's discuss in this thread.

**Comment by ymd-h** [comment]
@keiohta  Thank you for your request!

We might reduce data size by saving internal data structure directly, as long as user enables `next_of` or `stack_compress`.

However, saving internal data structure has problem of version compatibility.
Now, I only take care of public API of cpprb (not internal data structure).
If I provide save feature, users would expect their saved data can be used later. (Otherwise, they would be disappointed...)


Anyway, I need some investigation.
Do you know whether these overlapping data do affect compressed data size? (Are there any smart compression algorithm?)

**Reply by keiohta** [reply]
Thanks for considering the request!

I agree that directly saving internal data would be a good way to reduce data, and specifying `next_of` and `stack_compress` is reasonable effort for users who want to compress data more (as this feature is an **advanced** way to reduce data).

> Do you know whether these overlapping data do affect compressed data size? (Are there any smart compression algorithm?)

I'm afraid not, and just for clarification, I meant we can remove overlapped data (if an episode finishes with 1000 steps, then 999 states are overlapped), not meant it could improve compression rate due to the similarity between states and next_states.

> However, saving internal data structure has problem of version compatibility.

Does this mean the internal data structure can change in the future?

**Reply by ymd-h** [reply]
@keiohta  Please see my quick experiment.

The graphs show the compressed output file size for two situation.
One is "data" (blue) which has overlapping data like "obs" and "next_obs".
The other is "opt_data" (orange) which has no duplicated data.

The point is following. Except ".bz2", removing overlapping data doesn't change output data size, because compression algorithms are enough smart to detect overlapping data.
(This is the random value case and no other data are overlapped each other, so that compression level didn't improve the size.)

<img width="905" alt="SS 2021-03-17 22 27 19" src="https://user-images.githubusercontent.com/3405876/111475395-5cec2980-8770-11eb-9c42-6f1d2d7a0c1d.png">


```python
import os
import joblib
import numpy as np

import matplotlib.pyplot as plt

N = 1000

opt_obs = np.random.rand(N)
obs = opt_obs[:-1].copy()
next_obs = opt_obs[1:].copy()

opt_data = {"obs+next_obs": opt_obs}
data = {"obs": obs, "next_obs": next_obs}

data_size = {}
opt_data_size = {}
for suffix in (".z", ".gz", ".bz2", ".xz", ".lzma"):
    data_size[suffix] = []
    opt_data_size[suffix] = []
    for level in range(10):
        joblib.dump(data, f"data-{level}{suffix}", compress=level)
        joblib.dump(opt_data, f"opt_data-{level}{suffix}", compress=level)

        data_size[suffix].append(os.path.getsize(f"data-{level}{suffix}"))
        opt_data_size[suffix].append(os.path.getsize(f"opt_data-{level}{suffix}"))



fig = plt.figure(figsize=(15,5))
for i, suffix in enumerate((".z", ".gz", ".bz2", ".xz", ".lzma")):
    fig.add_subplot(1,5,i+1)
    plt.plot(data_size[suffix],label=f"data", marker=".", linestyle=":")
    plt.plot(opt_data_size[suffix],label=f"opt_data", marker=".", linestyle=":")
    plt.title(suffix)
    plt.xlabel("compress-level")
    plt.legend(loc="lower center")
    plt.ylim(bottom=0, top=12*N)
plt.show()
```

**Reply by keiohta** [reply]
Thanks for testing. The result is a bit surprise to me.

I'm curious how different state shape and batch size can affect the result. For example, `state_shape=(200,)`, and `batch_size=1e6`, which are reasonable choices for RL training. Can you test with this setting?

**Reply by ymd-h** [reply]
@keiohta 

Thank you for your comment. Data size is a good point to investigate.

I did another experiment.
The result shows that compression does not work for large data even though compression level is max.
(I also tried level-3, and the result was similar.)

<img width="464" alt="SS 2021-03-18 7 42 35" src="https://user-images.githubusercontent.com/3405876/111548149-98f9ab80-87bd-11eb-8506-4666b2306d21.png">

```python
import os
import joblib
import numpy as np

import matplotlib.pyplot as plt


# Default compression
suffix = ".gz"
level = 9


x = [int(1e+3), int(1e+4), int(1e+5), int(1e+6), int(1e+7)]

data_size = []
opt_data_size = []

for N in x:
    opt_obs = np.random.rand(N)
    obs = opt_obs[:-1].copy()
    next_obs = opt_obs[1:].copy()

    opt_data = {"obs+next_obs": opt_obs}
    data = {"obs": obs, "next_obs": next_obs}

    fname = f"data-{level}{suffix}"
    joblib.dump(data, fname, compress=level)
    data_size.append(os.path.getsize(fname))
    os.remove(fname)

    fname = f"opt_data-{level}{suffix}"
    joblib.dump(opt_data, fname, compress=level)
    opt_data_size.append(os.path.getsize(fname))
    os.remove(fname)


plt.figure(figsize=(7,5))
plt.plot(x, data_size,label=f"data", marker=".", linestyle=":")
plt.plot(x, opt_data_size,label=f"opt_data", marker=".", linestyle=":")
plt.title(suffix + f": level-{level}")
plt.xlabel("Original Data Size")
plt.legend()
plt.xscale("log")
plt.ylim(bottom=0, top=16.5*N)
plt.show()
```

**Reply by keiohta** [reply]
Thanks for checking it!
I think the surprising results suggest using the `opt_data` can further compress data and thus will be very helpful for users. Can you officially provide this feature?

**Reply by ymd-h** [reply]
@keiohta 

I add more comprehensive result. All algorithms work somehow similar.

I agree we can improve data size by removing overlapping region.
(However, I still don't know the best way.)

<img width="917" alt="SS 2021-03-18 8 00 05" src="https://user-images.githubusercontent.com/3405876/111549697-fc84d880-87bf-11eb-9c25-457ca71d37d2.png">

```python
import os
import joblib
import numpy as np

import matplotlib.pyplot as plt


# Default compression
level = 9


x = [int(1e+3), int(1e+4), int(1e+5), int(1e+6), int(1e+7)]

data_size = {}
opt_data_size = {}
for suffix in (".z", ".gz", ".bz2", ".xz", ".lzma"):
    data_size[suffix] = []
    opt_data_size[suffix] = []


    for N in x:
        opt_obs = np.random.rand(N)
        obs = opt_obs[:-1].copy()
        next_obs = opt_obs[1:].copy()

        opt_data = {"obs+next_obs": opt_obs}
        data = {"obs": obs, "next_obs": next_obs}

        fname = f"data-{level}{suffix}"
        joblib.dump(data, fname, compress=level)
        data_size[suffix].append(os.path.getsize(fname))
        os.remove(fname)

        fname = f"opt_data-{level}{suffix}"
        joblib.dump(opt_data, fname, compress=level)
        opt_data_size[suffix].append(os.path.getsize(fname))
        os.remove(fname)


plt.figure(figsize=(15,5))
fig = plt.figure(figsize=(15,5))
for i, suffix in enumerate((".z", ".gz", ".bz2", ".xz", ".lzma")):
    fig.add_subplot(1,5,i+1)
    plt.plot(x, data_size[suffix],label=f"data", marker=".", linestyle=":")
    plt.plot(x, opt_data_size[suffix],label=f"opt_data", marker=".", linestyle=":")
    plt.title(suffix + f": level-{level}")
    plt.xlabel("Original Data Size")
    plt.legend()
    plt.xscale("log")
    plt.ylim(bottom=0, top=16.5*N)
plt.show()
```

**Reply by keiohta** [reply]
Great. (btw, it might be better to show Y-axis with log-scale.)

Removing the overlapped data requires us to know the end of an episode, and I think we can do that by calling `on_episode_end` method similar to multi-step return. Although I'm not familiar with the internal structure, I think the combination of the following two realize the function.

1. specifying `next_of` and `stack_compress` when instantiating a buffer
2. calling `on_episode_end` when finishing an episode

How do you think of this?

**Reply by ymd-h** [reply]
@keiohta 

> 1. specifying next_of and stack_compress when instantiating a buffer
> 1. calling on_episode_end when finishing an episode

I agree, otherwise buffer cannot know.

In my quick idea, `def save(self, filename, *, compress=3, safe=True)` is a candidate API.

When user manually specify`safe=False`, then the buffer saves internal compressed data.
The default `safe=True` calls `get_all_transitions()`.

The other point is, can we depend joblib? (or any other library?)

**Reply by keiohta** [reply]
Yeah, I think that would be a good interface.

Another important point is how users can restore the compressed data. I think it would be great if API can also prepare a method that restore the compressed data and return exactly the same format with `get_all_transitions`, i.e., it includes `states` and `next_states`.

**Reply by keiohta** [reply]
> The other point is, can we depend joblib? (or any other library?)

I don't have knowledge around compression libraries... I think just using `joblib` would be fine.

**Reply by ymd-h** [reply]
@keiohta 

I released cpprb v10.1.0 just now.

`ReplayBuffer.save_transitions(file, *, safe=True)` and `ReplayBuffer.load_transitions(file)` are introduced.

If `next_of` and/or `stack_compress` are enabled **and** `safe=False` is manually specified, then buffer tries to reduce the file size.

Please try it!

**Comment by ymd-h** [comment]
Save and load transitions are introduced at cpprb v10.1.0

Please see [document](https://ymd_h.gitlab.io/cpprb/features/save_load_transitions/).
